lower.ts

```diff
@@ -139,6 +139,25 @@ function flattenStmt(s : AST.Stmt<Type>, blocks: Array<IR.BasicBlock<Type>>, env
         });
       return [...oinits, ...ninits];
     }
+
+    case "index-assign": {
+      var [oinits, ostmts, oval] = flattenExprToVal(s.obj, env);
+      var [iinits, istmts, ival] = flattenExprToVal(s.index, env);
+      var [ninits, nstmts, nval] = flattenExprToVal(s.value, env);
+
+      if (s.obj.a.tag !== "list") {
+        throw new Error("Compiler's cursed, go home.");
+      }
+      pushStmtsToLastBlock(blocks, ...ostmts, ...istmts, ...nstmts, {
```

Comment:
Here, for example, I think you could add a few extra statements to check that the index is in bounds and that the list is not none, and then generate a `store` that doesn't require any changes at all to the IR.

Reply:
We moved checking index in bound to lower from compiler

Reply:
Code was fixed to reflect this comment and the next.
We've moved all low-level code generation to lower.ts.